src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ResourceDataNode.cs

```diff
@@ -0,0 +1,191 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.Text;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    /// <summary>
+    /// Blob of data containing resources for all assemblies generated into the image.
+    /// Resources are simply copied from the inputs and concatenated into this blob.
+    /// All format information is provided by <see cref="ResourceIndexNode"/>
+    /// </summary>
+    internal class ResourceDataNode : ObjectNode, ISymbolNode
+    {
+        /// <summary>
+        /// Resource index information generated while extracting resources into the data blob
+        /// </summary>
+        private List<ResourceIndexData> IndexData { get; set; }
+        private int TotalLength { get; set; }
+
+        public ResourceDataNode()
+        {
+            _endSymbol = new ObjectAndOffsetSymbolNode(this, 0, "__embedded_resourcedata_End", true);
+        }
+
+        private ObjectAndOffsetSymbolNode _endSymbol;
+        public ISymbolNode EndSymbol => _endSymbol;
+
+        public override bool IsShareable => false;
+
+        public override ObjectNodeSection Section => ObjectNodeSection.ReadOnlyDataSection;
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        public int Offset => 0;
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix).Append("__embedded_resourcedata");
+        }
+
+        protected override string GetName() => this.GetMangledName();
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            // This node has no relocations.
+            if (relocsOnly)
+                return new ObjectData(Array.Empty<byte>(), Array.Empty<Relocation>(), 1, new ISymbolNode[] { this });
+            
+            byte[] blob = GenerateResourceBlob(factory);
+            return new ObjectData(
+                blob,
+                Array.Empty<Relocation>(),
+                1,
+                new ISymbolNode[]
+                {
+                                this,
+                                EndSymbol
+                });
+        }
+
+        public List<ResourceIndexData> GetOrCreateIndexData(NodeFactory factory)
```

Comment:
Nit: returning IEnumerable would be enough, so that nobody gets ideas about sneaking in new entries.